### PR TITLE
Python: Aggregate token usage across tool-call loop iterations in invoke_agent span

### LINKED ISSUE #4062: Python: [Bug]: invoke_agent span reports last LLM call's token usage instead of aggregate total
### Description

The `invoke_agent` span's `gen_ai.usage.input_tokens` and `gen_ai.usage.output_tokens` attributes only reflect the token  usage from the final `chat` completion call, rather than the sum of all LLM calls made during the agent's execution loop.

### Reproduction                                                                                                            

Run an agent with tool calling that requires multiple LLM round-trips (e.g., agent calls a tool, processes the result, then returns a final answer — 2 LLM calls total).

  ### Observed behavior

  | Span | `input_tokens` | `output_tokens` | Note |
  |------|---------------|-----------------|------|
  | Span 1 (`chat`) | 2,239 | 192 | LLM call 1 |
  | Span 2 (`execute_tool`) | — | — | Tool execution |
  | Span 3 (`chat`) | 2,569 | 99 | LLM call 2 |
  | **Span 4 (`invoke_agent`)** | **2,569** | **99** | **Same as Span 3 (last call only)** |

  ### Expected behavior

  | Span | `input_tokens` | `output_tokens` | Note |
  |------|---------------|-----------------|------|
  | **Span 4 (`invoke_agent`)** | **4,808** | **291** | **Sum of all `chat` spans** |

The `invoke_agent` span should aggregate token usage across all `chat` completion calls within the agent loop, so it  reflects the true total cost of the agent invocation.

### Impact

Observability dashboards and cost tracking tools that rely on the top-level `invoke_agent` span will undercount token usage for any agent execution that involves tool calling (multiple LLM round-trips). Users must manually sum child `chat` spans as a workaround.

### Environment

  - `agent-framework` version `1.0.0b260212`

### Code Sample

```markdown

```

### Error Messages / Stack Traces

```markdown

```

### Package Versions

agent-framework version 1.0.0b260212

### Python Version

_No response_

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by eavanvalkenburg**
### Motivation and Context

When an agent performs multiple LLM round-trips during function invocation (e.g., tool calls followed by a final response), the `invoke_agent` telemetry span only reported token usage from the last chat completion instead of the cumulative total across all iterations. This caused misleading observability data for cost tracking and debugging.

Fixes #4062

### Description

The root cause was in `FunctionInvocationLayer._get_response()` in `_tools.py`, which ran multiple `super_get_response` calls in a loop but returned only the last `ChatResponse` without accumulating `usage_details`. The fix introduces an `aggregated_usage` variable that uses the existing `add_usage_details()` helper to sum token counts after each chat completion, then assigns the aggregate back to `response.usage_details` before returning. Two regression tests verify correct aggregation for multi-call tool invocation and single-call scenarios.

### Contribution Checklist

- [X] The code builds clean without any errors or warnings
- [X] The PR follows the [Contribution Guidelines](https://github.com/microsoft/agent-framework/blob/main/CONTRIBUTING.md)
- [X] All unit tests pass, and I have added new tests where possible
- [ ] **Is this a breaking change?** If yes, add "[BREAKING]" prefix to the title of the PR.

---

> Note: PR autogenerated by eavanvalkenburg's agent

### CODE REVIEW

**Comment by markwallace-microsoft** [comment]
<a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/README.md">Python Test Coverage</a><details><summary>Python Test Coverage Report • </summary><table><tr><th>File</th><th>Stmts</th><th>Miss</th><th>Cover</th><th>Missing</th></tr><tbody><tr><td colspan="5"><b>packages/core/agent_framework</b></td></tr><tr><td>&nbsp; &nbsp;<a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py">_tools.py</a></td><td>934</td><td>83</td><td>91%</td><td><a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L186-L187">186&ndash;187</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L362">362</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L364">364</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L377">377</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L406-L408">406&ndash;408</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L416">416</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L434">434</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L448">448</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L455">455</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L462">462</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L479">479</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L481">481</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L488">488</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L496">496</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L531">531</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L580">580</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L584">584</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L627-L629">627&ndash;629</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L631">631</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L637">637</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L682-L684">682&ndash;684</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L686">686</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L709">709</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L735">735</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L777-L779">777&ndash;779</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L783">783</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L805">805</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L917-L923">917&ndash;923</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L959">959</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L971">971</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L973">973</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L975">975</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L978-L981">978&ndash;981</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1002">1002</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1006">1006</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1010">1010</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1024-L1026">1024&ndash;1026</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1367">1367</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1389">1389</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1476-L1482">1476&ndash;1482</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1611">1611</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1615">1615</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1661">1661</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1780">1780</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1810">1810</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1830">1830</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1832">1832</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1888">1888</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L1951">1951</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L2181-L2182">2181&ndash;2182</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L2237">2237</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L2310-L2311">2310&ndash;2311</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L2373">2373</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L2378">2378</a>, <a href="https://github.com/microsoft/agent-framework/blob/dc63729478d9b2dd342732313b3401e0b8bcd3a3/packages/core/agent_framework/_tools.py#L2385">2385</a></td></tr><tr><td><b>TOTAL</b></td><td><b>24011</b></td><td><b>2638</b></td><td><b>89%</b></td><td>&nbsp;</td></tr></tbody></table></details>

## Python Unit Test Overview
| Tests | Skipped | Failures | Errors | Time |
| ----- | ------- | -------- | -------- | ------------------ |
| 5239 | 20 :zzz: | 0 :x: | 0 :fire: | 1m 23s :stopwatch: |

**Review by moonbox3 — COMMENTED** [review]
## Automated Code Review

**Reviewers:** 4 | **Confidence:** 89%

### ✓ Correctness
> The diff correctly adds usage aggregation across multiple LLM round-trips in FunctionInvocationLayer's non-streaming path. The add_usage_details helper is correctly used: it handles None inputs safely, creates new dicts when summing, and the aggregated usage is properly set on the final ChatResponse before returning. All exit paths in the loop ("return" action, loop exhaustion + final call) correctly aggregate and assign usage. The ChatTelemetryLayer captures individual per-call usage in its own spans before FunctionInvocationLayer mutates the response, so individual chat spans retain their original values. Tests are well-structured and validate both multi-call and single-call scenarios correctly.

### ✓ Security Reliability
> The diff adds token usage aggregation across multiple LLM round-trips in the function invocation loop. The `add_usage_details` helper correctly handles `None` inputs for both arguments, and the aggregated usage is properly assigned back to the response before returning. The `UsageDetails` type is a TypedDict with integer fields, so there are no injection or deserialization risks. The test coverage exercises both the multi-call and single-call paths. No security or reliability concerns found.

### ✗ Test Coverage
> The diff adds usage aggregation across multiple LLM calls in the non-streaming function invocation loop and includes two well-structured tests: one for multi-round-trip tool calls and one for a single call. Both tests make meaningful assertions on both individual chat spans and the aggregate invoke_agent span. However, test coverage has gaps: there is no test exercising the max-iterations/loop-exhaustion code path (diff lines 2268-2269) where aggregation also occurs, no test for the case where one or more ChatResponse objects return usage_details=None, and no test covering the streaming path (which notably does NOT receive the aggregation logic — this may be intentional but is worth a test to document expected behavior).

### ✓ Design Approach
> The fix correctly addresses a real bug where only the last LLM call's `usage_details` was surfaced on the returned `ChatResponse` when the function-invocation loop made multiple round-trips. The accumulation logic is sound: usage is added to `aggregated_usage` immediately after each `super_get_response` call, and is stamped onto the response before every exit point (early return inside the loop, and the post-loop fallback call). The streaming path does not need the same fix because it yields `UsageContent` items from every inner stream, which `ChatResponse.from_updates` → `_process_update` already sums correctly. One minor structural issue exists: the `from ._types import add_usage_details` import is placed inside the nested `_get_response()` closure rather than alongside the existing deferred `from ._types import (...)` block at the top of the outer `get_response()` method, where it would be more consistent with the surrounding code style.

### Flagged Issues
- [ ] No test covers the loop-exhaustion / max_iterations code path (diff lines ~2268-2269) where usage is also aggregated. This is a distinct branch in _get_response that exits after the for-loop with tool_choice='none', and the aggregation there is completely untested.

### Suggestions
- Add a test where one or more ChatResponse objects have usage_details=None to verify graceful handling through the aggregation path at the integration level.
- Consider adding a test confirming that streaming responses do not aggregate usage separately (since the streaming path already sums via UsageContent items), to document the expected asymmetry between streaming and non-streaming paths.
- When add_usage_details(None, usage2) is called it returns the same usage2 object rather than a copy. This is harmless today because aggregated_usage is never mutated in-place, but a defensive copy (e.g., UsageDetails(**usage2)) would prevent subtle aliasing bugs if future code changes that assumption.
- Move `from ._types import add_usage_details` out of the nested `_get_response()` closure and into the existing `from ._types import (ChatResponse, ChatResponseUpdate, ResponseStream)` block at the top of `get_response()`, for consistency with the surrounding deferred-import pattern.

---
*Automated review by moonbox3's agents*

File: python/packages/core/tests/core/test_observability.py [hunk 1]
```diff
@@ -2781,3 +2782,94 @@ def mock_get_meter(*args, **kwargs):
...
+
+    span_exporter.clear()
+    await agent.run(messages="Hi")
+
+    spans = span_exporter.get_finished_spans()
+    invoke_spans = [s for s in spans if s.attributes.get(OtelAttr.OPERATION.value) == OtelAttr.AGENT_INVOKE_OPERATION]
+    assert len(invoke_spans) == 1
+
+    assert invoke_spans[0].attributes.get(OtelAttr.INPUT_TOKENS) == 100
+    assert invoke_spans[0].attributes.get(OtelAttr.OUTPUT_TOKENS) == 50
```
Comment (moonbox3) [inline python/packages/core/tests/core/test_observability.py:2875]: This test only covers the early-return path (action=='return'). The diff also adds aggregation at lines ~2268-2269 for the loop-exhaustion path (max_iterations exceeded → final tool_choice='none' call). Add a test that triggers max_iterations exhaustion (e.g., set max_iterations=1 with a tool call that keeps looping) to cover the second aggregation site.

File: python/packages/core/agent_framework/_tools.py [hunk 2]
```diff
@@ -2152,6 +2153,8 @@ def get_response(
         if not stream:
```
Comment (moonbox3) [inline python/packages/core/agent_framework/_tools.py:2155]: This import belongs alongside the existing `from ._types import (ChatResponse, ChatResponseUpdate, ResponseStream)` block at the top of `get_response()` (around line 1848). Placing it inside the nested async closure is inconsistent with the established pattern.

```suggestion
        from ._types import (
            ChatResponse,
            ChatResponseUpdate,
            ResponseStream,
            add_usage_details,
        )
```

File: python/packages/core/tests/core/test_observability.py (hunk 1 above)
Comment (eavanvalkenburg) [reply python/packages/core/tests/core/test_observability.py:2875]: Fixed in `dc637294`.

File: python/packages/core/agent_framework/_tools.py (hunk 2 above)
Comment (eavanvalkenburg) [reply python/packages/core/agent_framework/_tools.py:2155]: Fixed in `dc637294`.

**Review by moonbox3 — APPROVED** [review]

**Review by giles17 — APPROVED** [review]

**Comment by 0x7c13** [comment]
Oh, I thought this was intended behavior vs the correct total aggregated tokens logic for streaming path. Should have reported sooner :) I thought we should manually calculate the total earlier :(
